File: algorithms/lua/fft.lua
local math = math

-- complex ops با table
local function cadd(a, b) return {a[1]+b[1], a[2]+b[2]} end
local function csub(a, b) return {a[1]-b[1], a[2]-b[2]} end
local function cmul(a, b)
    return {a[1]*b[1]-a[2]*b[2], a[1]*b[2]+a[2]*b[1]}
end
local function cabs(a) return math.sqrt(a[1]*a[1] + a[2]*a[2]) end

local function fft(a, n)
    -- bit-reversal
    local j = 0
    for i = 1, n-1 do
        local bit = n >> 1
        while j & bit ~= 0 do j = j ~ bit; bit = bit >> 1 end
        j = j ~ bit
        if i < j then a[i+1], a[j+1] = a[j+1], a[i+1] end
    end
    -- butterfly
    local len = 2
    while len <= n do
        local ang  = 2 * math.pi / len
        local wlen = {math.cos(ang), math.sin(ang)}
        local i = 0
        while i < n do
            local w = {1.0, 0.0}
            for jj = 0, len//2-1 do
                local u = a[i+jj+1]
                local v = cmul(a[i+jj+len//2+1], w)
                a[i+jj+1]       = cadd(u, v)
                a[i+jj+len//2+1]= csub(u, v)
                w = cmul(w, wlen)
            end
            i = i + len
        end
        len = len << 1
    end
end

local function main()
    local n   = tonumber(io.read("*l"))
    local a   = {}
    local idx = 1
    for v in io.read("*l"):gmatch("%S+") do
        a[idx] = {tonumber(v), 0.0}
        idx = idx + 1
    end

    local start = os.clock()
    fft(a, n)
    local ms = (os.clock() - start) * 1000

    local k   = math.min(5, n)
    local sum = 0
    for i = 1, k do sum = sum + cabs(a[i]) end

    print(string.format("%.6f", sum))
    io.stderr:write(string.format("%.3f\n", ms))
end

main()

File: algorithms/c/fft.c
#include <stdio.h>
#include <stdlib.h>
#include <math.h>
#include <time.h>

#ifndef M_PI
#define M_PI 3.14159265358979323846
#endif

typedef struct { double r, i; } Complex;

static inline Complex cadd(Complex a, Complex b){ return (Complex){a.r+b.r, a.i+b.i}; }
static inline Complex csub(Complex a, Complex b){ return (Complex){a.r-b.r, a.i-b.i}; }
static inline Complex cmul(Complex a, Complex b){
    return (Complex){a.r*b.r - a.i*b.i, a.r*b.i + a.i*b.r};
}
static inline double  cabs2(Complex a){ return sqrt(a.r*a.r + a.i*a.i); }

void fft(Complex *a, int n) {
    /* bit-reversal */
    for (int i = 1, j = 0; i < n; i++) {
        int bit = n >> 1;
        for (; j & bit; bit >>= 1) j ^= bit;
        j ^= bit;
        if (i < j) { Complex t = a[i]; a[i] = a[j]; a[j] = t; }
    }
    /* butterfly */
    for (int len = 2; len <= n; len <<= 1) {
        double ang  = 2.0 * M_PI / len;
        Complex wlen = {cos(ang), sin(ang)};
        for (int i = 0; i < n; i += len) {
            Complex w = {1.0, 0.0};
            for (int j = 0; j < len / 2; j++) {
                Complex u = a[i+j];
                Complex v = cmul(a[i+j+len/2], w);
                a[i+j]         = cadd(u, v);
                a[i+j+len/2]   = csub(u, v);
                w = cmul(w, wlen);
            }
        }
    }
}

int main() {
    int n;
    scanf("%d", &n);
    Complex *a = malloc(n * sizeof(Complex));
    for (int i = 0; i < n; i++) { scanf("%lf", &a[i].r); a[i].i = 0.0; }

    struct timespec start, end;
    clock_gettime(CLOCK_MONOTONIC, &start);
    fft(a, n);
    clock_gettime(CLOCK_MONOTONIC, &end);

    double ms = (end.tv_sec  - start.tv_sec)  * 1000.0
              + (end.tv_nsec - start.tv_nsec) / 1e6;

    int    k   = n < 5 ? n : 5;
    double sum = 0.0;
    for (int i = 0; i < k; i++) sum += cabs2(a[i]);
    printf("%.6f\n", sum);
    fprintf(stderr, "%.3f\n", ms);

    free(a);
    return 0;
}